src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,71 +9,347 @@
     MPIR_STATUS_SET_COUNT(rreq->status, recv_data_sz);
     rreq->status.MPI_SOURCE = MPIDIG_REQUEST(rreq, rank);
     rreq->status.MPI_TAG = MPIDIG_REQUEST(rreq, tag);
+    MPIDI_XPMEM_REQUEST(rreq, counter_ptr) = NULL;
 
-    /* Send ack to sender */
-    slmt_ack_hdr->sreq_ptr = sreq_ptr;
     mpi_errno =
-        MPIDI_SHM_do_ctrl_send(MPIDIG_REQUEST(rreq, rank),
-                               MPIDIG_context_id_to_comm(MPIDIG_REQUEST(rreq, context_id)),
-                               MPIDI_SHM_XPMEM_SEND_LMT_ACK, &ack_ctrl_hdr);
+        MPIDI_XPMEM_seg_regist(src_lrank, src_data_sz, (void *) src_offset, &seg_ptr, &src_buf,
+                               MPIDI_XPMEM_USER_BUFFER);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    XPMEM_TRACE("handle_lmt_single_recv: handle matched rreq %p [source %d, tag %d, "
+                " context_id 0x%x], copy dst %p, src %lx, bytes %ld\n", rreq,
+                MPIDIG_REQUEST(rreq, rank), MPIDIG_REQUEST(rreq, tag),
+                MPIDIG_REQUEST(rreq, context_id), (char *) MPIDIG_REQUEST(rreq, buffer),
+                src_offset, recv_data_sz);
+
+    /* Copy data to receive buffer */
+    mpi_errno = MPIR_Localcopy(src_buf, recv_data_sz, MPI_BYTE,
+                               MPIDIG_REQUEST(rreq, buffer), MPIDIG_REQUEST(rreq, count),
+                               MPIDIG_REQUEST(rreq, datatype));
+    MPIR_ERR_CHECK(mpi_errno);
+
+    ack_ctrl_hdr.xpmem_slmt_send_fin.req_ptr = sreq_ptr;
+    mpi_errno = MPIDI_SHM_do_ctrl_send(MPIDIG_REQUEST(rreq, rank),
+                                       MPIDIG_context_id_to_comm(MPIDIG_REQUEST
+                                                                 (rreq, context_id)),
+                                       MPIDI_SHM_XPMEM_SEND_LMT_RECV_FIN, &ack_ctrl_hdr);
     MPIR_ERR_CHECK(mpi_errno);
 
     MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(rreq, datatype));
+    if (MPIDI_XPMEM_REQUEST(rreq, counter_ptr))
+        MPIR_Handle_obj_free(&MPIDI_XPMEM_cnt_mem, MPIDI_XPMEM_REQUEST(rreq, counter_ptr));
     MPID_Request_complete(rreq);
 
+    mpi_errno = MPIDI_XPMEM_seg_deregist(seg_ptr);
+    MPIR_ERR_CHECK(mpi_errno);
+
   fn_exit:
-    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_HANDLE_LMT_SINGLE_RECV);
     return mpi_errno;
   fn_fail:
     goto fn_exit;
 }
 
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_handle_lmt_recv(uint64_t src_offset, size_t data_sz,
+                                                         uint64_t sreq_ptr, int src_lrank,
+                                                         MPIR_Comm * comm, MPIR_Request * rreq)
+{
+
+    int mpi_errno = MPI_SUCCESS;
+    int recvtype_iscontig;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
+    /* TODO: need to support non-contig datatype on receiver side. */
+    MPIR_Datatype_is_contig(MPIDIG_REQUEST(rreq, datatype), &recvtype_iscontig);
```

Comment:
We support non-contig recv datatype but just use single-copy. Can you please improve the TODO comment?

Reply:
Do we want to use a larger threshold for coop copy? I am afraid that it does not perform better than single-copy for medium size messages.

Reply:
I need to measure and check what threshold is suitable

Reply:
I'd suggest add the new CVAR first, then you can measure it and adjust the value.